[Question Answering] Authoring Client Initialization

AhmedLeithy opened the pull request:
Hey all,

This PR is for initializing the QuestionAnswering authoring client. I have the following concerns:

1.  The client is generated with the name `QuestionAnsweringProjectsClient`, with a plural "projects". I think it makes sense to have the plural "projects" in the name of the client, and as such used the same for the namespace. Do we want the namespace to be `...QuestionAnswering.Project` instead?

2. Using the `namespace` parameter in autorest.md does not actually set the namespace of the generated file(s). Not sure if this is a bug or if this some sort of mistake on my part, but I changed the namespace using the [CodeGenType] attribute. I left the syntax that I was using as a comment in autorest.md for demonstration.

3. I removed the newly created options class as mentioned in the comment that was left in autorest.md. I would also like to note that we can keep the class but make it generate with a different name, `QuestionAnsweringProjectsClientOptions` for example. This can be done by adding another `Title` property in the autorest.md file for the generated swagger. This will be useful if we expect the runtime and authoring clients to have different version names and/or be out of sync with versioning.

AhmedLeithy committed:
Question Answering LLC Init

AhmedLeithy committed:
Removed QuestionAnsweringProjectsClientOptions

AhmedLeithy committed:
Exported API

azure-sdk commented:
API changes have been detected in `Azure.AI.Language.QuestionAnswering`. You can review API changes [here](https://apiview.dev/Assemblies/Review/d0db0b241e3249718696a7b18e241a71?diffOnly=True&diffRevisionId=ae1bde9350464af2a45eab9cc03c2857)

heaths commented:
1. Given the plural "Projects" is what the service team used for the operation ID, `QuestionAnsweringProjectsClient` sounds correct: https://github.com/Azure/azure-rest-api-specs/blob/f3c2b263da9afc93edb71d55d9b08c75de7401d9/specification/cognitiveservices/data-plane/Language/stable/2021-10-01/questionanswering-authoring.json#L33

   As for the subnamespace, I think "Projects" still makes sense given plurality of several other subnamespaces, but defer to @KrzysztofCwalina and @tg-msft, as well as @annatisch and @johanste if we want to have a subnamespace in Python.

2. Where did you put the `namespace` declaration in _autorest.md_? In the `batch`? This should've worked. Given how many models there are, I'd prefer we don't have to create and maintain a bunch of classes just to move. If anything, a transform to add `x-namespace` to each model would be concise:

   ```yaml
   directive:
   - from: swagger-document
     where: $.definitions.*
     transform: $["x-namespace"] = "Azure.AI.Language.QuestionAnswering.Projects";
   ```

3. Having a single `QuestionAnsweringClientOptions` makes sense. We've done that elsewhere when a separate class provides literally no value such as extra properties. We can always derive a specific class later if necessary.

heaths commented on sdk/cognitivelanguage/Azure.AI.Language.QuestionAnswering/src/Azure.AI.Language.QuestionAnswering.csproj:
```diff
@@ -7,6 +7,7 @@
     <ApiCompatVersion>1.0.0</ApiCompatVersion>
     <PackageTags>Azure AI Language Question Answering QnA Maker</PackageTags>
     <TargetFrameworks>$(RequiredTargetFrameworks)</TargetFrameworks>
+    <DefineConstants>$(DefineConstants);EXPERIMENTAL</DefineConstants>
```
What's `EXPERIMENTAL` for? I don't see it referenced in a browser-based search within this PR (with all file content rendered).

AhmedLeithy replied:
The effect of having this is that it activates experimental Azure.Core functions that are used by the LLC. Without it, I get errors about the following functions not being found:
 
- `LowLevelOperationHelpers.ProcessMessage()` + async version.
- An overload of the `_pipeline.ProcessMessage()` + async version.

I found this Constant being included in the [Azure.Template.LLC.csproj](https://github.com/Azure/azure-sdk-for-net/blob/117081fc19582e81a0f5be0f2607c77f83f0c9fe/sdk/template-LLC/Azure.Template.LLC/src/Azure.Template.LLC.csproj#L10). I also found it being included in a similar way in other SDK projects including Synapse, WebPubSub and AgriFood.

Is this not the correct solution?

heaths replied:
If that's what it takes, seems correct. There's no final documentation for LLC yet that I'm aware.

heaths commented on sdk/cognitivelanguage/Azure.AI.Language.QuestionAnswering/api/Azure.AI.Language.QuestionAnswering.netstandard2.0.cs:
```diff
@@ -188,6 +188,57 @@ public partial class TextDocument
+        public virtual System.Threading.Tasks.Task<Azure.Operation<System.BinaryData>> DeployProjectAsync(string projectName, string deploymentName, Azure.RequestContext context = null) { throw null; }
+        public virtual Azure.Operation<System.BinaryData> Export(string projectName, string format = null, string assetKind = null, Azure.RequestContext context = null) { throw null; }
+        public virtual System.Threading.Tasks.Task<Azure.Operation<System.BinaryData>> ExportAsync(string projectName, string format = null, string assetKind = null, Azure.RequestContext context = null) { throw null; }
+        public virtual Azure.Response GetDeleteStatus(string jobId, Azure.RequestContext context = null) { throw null; }
```
@annelo-msft @pakrym should raw status getters be added, or was this because the LRO method wasn't marked explicitly as an LRO? https://github.com/Azure/azure-rest-api-specs/blob/f3c2b263da9afc93edb71d55d9b08c75de7401d9/specification/cognitiveservices/data-plane/Language/stable/2021-10-01/questionanswering-authoring.json#L187-L194

Though, I do see methods like `GetExportStatus` below where that method is marked as an LRO. If that's expected, that's fine. I just thought maybe we'd generate an `Operation<T>` or is that considered too high-level?

/cc @rokulka @annatisch @mshaban-msft

heaths replied:
Still, @rokulka, any reason delete is not marked as an LRO while import, export, deploy, and update (sources, QnAs) are? Delete is returning an `Operation-Location` header. I also notice the job status endpoint is different from the others: https://github.com/Azure/azure-rest-api-specs/blob/f3c2b263da9afc93edb71d55d9b08c75de7401d9/specification/cognitiveservices/data-plane/Language/stable/2021-10-01/questionanswering-authoring.json#L209

heaths replied:
I ping @rokulka offline (with you and @annatisch) as well on this one. If it's a quick turnaround (fix or no fix), let's wait to get it in this PR; otherwise, if some time is required feel free to have this one merged, open a tracking issue on this potential issue, and we can get it in another PR.

rokulka replied:
@heaths  @annelo-msft  @pakrym - Project Delete is also an LRO. Sorry, I missed adding the property in the swagger - this API was a last minute addition.  The Operation-Location URL returned is for the path /deletion-jobs/{jobId} .

heaths replied:
What timeframe do you have for pushing a change? Otherwise, @AhmedLeithy you can use a swagger transform. Talk to @tg-msft in my stead if you need help.

johanste replied:
Let's make sure we update the swagger document if there is missing information in there. We should avoid transforms that only exist because we didn't get around to updating the swagger itself.

heaths replied:
@johanste agreed, but *temporary* transforms can be useful to unblock. We still need the swagger change, yes. But it looks like Azure/azure-rest-api-specs#16832 is close to getting merged to main.

heaths replied:
Offline, @AhmedLeithy said the changes were merged a while back and necessary changes herein.

heaths commented on sdk/cognitivelanguage/Azure.AI.Language.QuestionAnswering/src/QuestionAnsweringProjectsClient.cs:
```diff
@@ -0,0 +1,17 @@
+namespace Azure.AI.Language.QuestionAnswering.Projects
+{
+    /// <summary> The QuestionAnsweringProjects service client. </summary>
+    [CodeGenType("QuestionAnsweringProjectsClient")]
```
Despite my comment about the `namespace` directive in _autorest.md_, since we're not *currently* generating models for LLCs, this is probably fine. I had forgotten that we are not currently generating models, though I hear it's still being considered for .NET.

AhmedLeithy replied:
Sounds good. I will keep it as is for now, but if we end up generating models, I can try the transform solution you previously suggested.

And yea, I tried adding the namespace parameter in the batch file under the input-file for authoring (it's commented out in autorest.md now). Didn't work, not sure why.

AhmedLeithy replied:
Btw, is there a specific reason we're not generating models for this client? Why did we choose to go with an LLC for authoring?

heaths replied:
In general, the Azure SDK team is defaulting to LLCs for less-used APIs. We've done a bunch of user studies across different languages. It's more so a scalability problem (we can't scale to hand-author everything, even with great partners like you!), but the user research also showed some advantages. In mixed cases like Cognitive Services, the authoring APIs are less used than inference, and in some cases are rarely used even then when a great Azure Portal (or other portal, like Cognitive Language has) experience is available.

Across the languages, IIRC for all of them, we don't do models because customers often have their own serializable models or capabilities already. But we still generate clients because it uses the same auth, pipeline, logging, etc., and in some languages like with C# / .NET we generate method names as well for discoverability.

heaths commented on sdk/cognitivelanguage/Azure.AI.Language.QuestionAnswering/src/autorest.md:
```diff
@@ -13,9 +13,11 @@ batch:
-# - input-file: https://raw.githubusercontent.com/Azure/azure-rest-api-specs/34a2c0723155d134311419fd997925ce96b85bec/specification/cognitiveservices/data-plane/Language/stable/2021-10-01/questionanswering-authoring.json
-#   add-credentials: true
-#   low-level-client: true
+- input-file: https://raw.githubusercontent.com/Azure/azure-rest-api-specs/34a2c0723155d134311419fd997925ce96b85bec/specification/cognitiveservices/data-plane/Language/stable/2021-10-01/questionanswering-authoring.json
```
I just noticed that line 13 above (`model-namespace: false`) has too many spaces. It still seems to work, but could you fix the spacing and generate again just to make sure everything is technically correct?

AhmedLeithy replied:
Sorry but I don't seem to be able to find the extra spaces. Do you mean the indentation before `model-namespace: false`? Autorest does not generate without having two spaces before it like the rest of the properties. 

The first property for each swagger is the only one with a single space. i.e. `- input-file: ...`

heaths replied:
Hmm. When I looked at my local source I don't see extra spaces. I guess it's just a GitHub rendering issue. Never mind then.

heaths reviewed (changes requested):
Overall looks great, but a couple of open questions we should probably resolve.

AhmedLeithy committed:
added "Qnas" to cspell

AhmedLeithy committed:
Regenerated after Delete LRO fix

azure-sdk commented:
API changes have been detected in `Azure.AI.Language.QuestionAnswering`. You can review API changes [here](https://apiview.dev/Assemblies/Review/d0db0b241e3249718696a7b18e241a71?diffOnly=True&diffRevisionId=ae1bde9350464af2a45eab9cc03c2857)

AhmedLeithy committed:
Regenerated Client + Added reference to Azure.Core

AhmedLeithy committed:
Exported APIs

AhmedLeithy committed:
Added transforms to accomodate OperationId changes

AhmedLeithy commented:
As per @rokulka 's request and to be consistent with python, I've added transforms to ensure that the recent changes made to the operationIds in the swagger do not affect our SDK. I've also updated autorest.md to reference the latest swagger.

heaths commented on sdk/cognitivelanguage/Azure.AI.Language.QuestionAnswering/src/Azure.AI.Language.QuestionAnswering.csproj:
```diff
@@ -33,5 +33,8 @@
     <Compile Include="$(AzureCoreSharedSources)TaskExtensions.cs" LinkBase="Shared" />
   </ItemGroup>
+  <ItemGroup>
+    <ProjectReference Include="..\..\..\core\Azure.Core\src\Azure.Core.csproj" />
```
Azure.Core should almost always be a `PackageReference`.

AhmedLeithy replied:
When regenerating the client using the updated Autorest.Csharp package, the client references an overload of the methods 'CreateMessage' and 'WaitForCompletion' Async that are not available in the Core package. There doesn't seem to be any updates to the Core package.

As such, the project does not build without this project reference. I believe the alternative would be to wait for a new Core package (I'm not sure how frequently it's released). 

By the way, I have included you in an email thread about a bug that prompted me to update Autorest.Csharp package in the first place.

heaths replied:
@pakrym what should the process be in this case? We've had a few over the past couple years so it's a bit of a jumble what's now correct.

pakrym replied:
cc @annelo-msft

annelo-msft replied:
@heaths, when we make changes to Azure.Core that are required by the client libraries, we move the libraries over to a project reference so they'll continue to build until the new Azure.Core package is released.  Once Azure.Core is released (usually a few days before the client library package releases), we change those back to package references.

heaths replied:
Didn't we have some properties that some Azure.Core.props file used to do that, or something like that? I don't see the file now, though. If this indeed is the consistent way I can sign off, but I know we had a different way of doing this once upon a time.

At the very least, @AhmedLeithy you should open a bug and add a TODO or (old school) BUGBUG comment referencing it to make sure it gets cleaned up. Accidentally shipping this SDK against a ProjectReference of Azure.Core can and has (for another SDK) caused issues.

AhmedLeithy replied:
I've added a TODO comment and created [this Bug/Issue](https://github.com/Azure/azure-sdk-for-net/issues/26171) to remind us to change it before release:

heaths commented on sdk/cognitivelanguage/Azure.AI.Language.sln:
```diff
@@ -4,6 +4,9 @@ Microsoft Visual Studio Solution File, Format Version 12.00
 VisualStudioVersion = 17.0.31815.197
 MinimumVisualStudioVersion = 10.0.40219.1
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Azure.AI.Language.QuestionAnswering", "Azure.AI.Language.QuestionAnswering\src\Azure.AI.Language.QuestionAnswering.csproj", "{9507A18C-F21A-4305-80B8-8E9DB4E5D351}"
+	ProjectSection(ProjectDependencies) = postProject
+		{8DAB9728-0F12-40DF-A5B7-EFD602AF3C0E} = {8DAB9728-0F12-40DF-A5B7-EFD602AF3C0E}
+	EndProjectSection
```
What does this do? If not necessary, I'd back this change out.

AhmedLeithy replied:
I believe this adds a project dependency for QA on the Azure.Core project. It was automatically added when adding the reference the project reference. Should I still remove it?

heaths replied:
No, assuming we keep the ProjectReference, but I'm surprised a new project wasn't added to the SLN itself. That's why this diff block confused me.

AhmedLeithy committed:
Added TODO comment

heaths reviewed (approved):
I'd make a tracking issue and reference it in the comment, but otherwise LGTM.
